release/tasks/calculate-new-versions.ts

```diff
@@ -0,0 +1,76 @@
+import { Context, Bump, Tag, Package } from '../@types/custom';
+import * as semver from 'semver';
+
+import { debug } from '../lib/utils';
+import updateDependencies from '../lib/update-dependencies';
+
+/**
+ *
+ * @param tag The tag of the commit
+ */
+const commitType = (tag: Tag): Bump => {
+
+    switch (tag) {
+        case 'Docs': return Bump.none;
+        case 'Build': return Bump.none;
+        case 'Upgrade': return Bump.patch;
+        case 'Chore': return Bump.patch;
+        case 'Fix': return Bump.patch;
+        case 'New': return Bump.minor;
+        case 'Breaking': return Bump.major;
+        default: return Bump.none;
+    }
+};
+
+
+const calculatePackageNewVersion = (pkg: Package, bump: Bump): string => {
+
+    if (!pkg.publishedVersion) {
+        return pkg.content.newVersion;
+    }
+
+    const newVersion = semver.inc(pkg.oldVersion, Bump[bump] as semver.ReleaseType)!;
+
+    debug(`Bumping ${pkg.name} from ${pkg.oldVersion} to ${newVersion}`);
+
+    return newVersion;
+};
+
+/**
+ * Bumps the version of each package:
+ *
+ * 1. Using the commit history of each package calculates the new version bump
+ * 1. Updates the inter references of each package
+ * 1. Updates the versions of all the packages (`patch`) that do not have any commits but
+ *    have changed versions in their dependencies
+ */
+export const calculateNewVersions = (ctx: Context) => {
+
+    const { packages } = ctx;
+
+    // Step 1
+    packages.forEach((pkg) => {
+
+        const bumpType = pkg.commits.reduce((currentBump, commit) => {
+            const bump = commitType(commit.tag);
+
+            return Math.max(currentBump, bump);
+        }, Bump.none);
+
+        if (bumpType === Bump.none) {
+            return;
+        }
+
+        pkg.content.version = calculatePackageNewVersion(pkg, bumpType);
+    });
+
+    // Step 2
+    updateDependencies(ctx);
+
+    // Step 3
```

Comment:
Probably also worth repeating what each step does here (even though it's in the comment for the whole function). It took me a second to get my bearings on the first read.